### DISCUSSION TITLE
0.9.3 - LittleFS, Multiple Compilation Units and more FFT

### DISCUSSION THREAD

**Discussion by pschatzmann** [pschatzmann/arduino-audio-tools#466] [Announcements]
- Support for LIttleFS for AudioPlayer
- Support for cmsis fft
- Support for FFT using the Espressif DSP library
- added FFT tests
- A2DPStream remove instance() method
- SineFromTable performance optimisations
- added FastSineGenerator class (thanks to[vivichrist](https://github.com/vivichrist)
- STM32 PWM output
- Examples for using SD library
- Guitar Effects example using Web GUI
- Support for multiple compilation units (with USE_INLINE_VARS in AudiConfig.h)
- Example for using FFT from A2DP
- Some other smaller error corrections

<hr /><em>This discussion was created from the release <a href='https://github.com/pschatzmann/arduino-audio-tools/releases/tag/v0.9.3'>0.9.3 - LittleFS, Multiple Compilation Units and more FFT</a>.</em>

**Comment by John2009R** [comment]
Nice. For the guitar effects example, which library provides "HttpServer.h"?

**Reply by pschatzmann** [reply]
I was using my TinyHttp Library: see my [Blogs](https://www.pschatzmann.ch/home/2022/10/14/guitar-effects-with-web-gui/)

**Reply by John2009R** [reply]
Oh I see now. I tried it before your last commit (before you commented the include files).
It worked for me and looks like there is a lot of potential. I was worried what the latency would be like but I think it is fine.

Anyone else trying this make sure you do it with headphones if you have a board with the mics and line in connected in parallel. I got a big feedback squeal.

**Reply by pschatzmann** [reply]
Yes I added this information based on your comment.
So they still did not correct this microphone hardware bug in your board version as well. 
I gave up any hope that AIThinker will ever correct this bug...

**Reply by John2009R** [reply]
The ESP32-A1S has separate inputs for Mic and line and on my board I can see the traces for line going straight to those pins (via caps). 
The Mics appear to go to the other side. If this is a bug it is probably in the ESP32-A1S itself not the pcb, or a configuration setting in the es8388.

**Reply by pschatzmann** [reply]
The board has been designed for the AC101 and there it is working fine. Then they just replaced the AC101 with an ES8388 which works a little bit different w/o adapting the board design. 
The LyraT is working properly!
Unfortunately this problem can not be resolved with software...
Further info can be found [here](https://www.pschatzmann.ch/home/2021/12/15/the-ai-thinker-audiokit-audio-input-bug/)
